### PR TITLE
kubeadm: Make self-hosting work and split it out to a phase

### LINKED ISSUE #127: Follow-up selfhosting improvements
https://github.com/kubernetes/kubernetes/pull/40075 is now lgtm'd and we don't want to put more complexity in the PR, it's big enough already and has been waiting for quite a long time.

It now works, and that's great.

~~OLD: Here we should track the fixup work (20th Jan):~~
 - ~~Be able to deploy 2 schedulers and cms~~
 - ~~Specify rollingupdate parameters to the deployments: https://github.com/kubernetes/kubernetes/pull/40075#discussion_r97107015~~
 - ~~Use server-side DaemonSet upgrades - depends on https://github.com/kubernetes/kubernetes/pull/41116~~
 - ~~General fixup of the code/make it more efficient~~
 - ~~Add tests~~
 - ~~Refactor the API type behind it so we can specify more options (like 2 or more schedulers)~~
 - ~~Make secrets of the needed certs and mount them in the Pod~~
 - ~~Use the more generic `--controlplane` or `--controlplane-type` flag~~
 - ~~Make constants of commonly used string values~~
 - ~~Adopt the Pod checkpointer https://github.com/kubernetes/kubeadm/issues/131~~
 ...and probably more. 

**EDIT: New list as of 19th June**
 - [x] Make self-hosting work, split out to phase, use DaemonSets: https://github.com/kubernetes/kubernetes/pull/47435
 - [x] Upload certificates to the API using Secrets: https://github.com/kubernetes/kubernetes/pull/42548
 - [ ] Propose a create-first-then-delete DaemonSetUpdateStrategy and consume for the controller-manager and scheduler
 - [ ] Write a kubelet checkpointing proposal and opt-into that for kubeadm

@pires @mikedanese @kargakis @lukemarsden

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

 - Removes the old self-hosting code
 - Puts the new self-hosting code in `phases/selfhosting`
   - The new code reads manifests from disk (static pods)...
   - ...mutates the PodSpec as necessary...
   - ...and posts the DaemonSet to the API Server...
   - ...and waits for it to come up
 - Uses DaemonSets for all control plane components
 - Creates a `kubeadm alpha phase selfhosting` command that can be invoked against any kubeadm-cluster after install.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes: https://github.com/kubernetes/kubeadm/issues/127
(large part of at least)

**Special notes for your reviewer**:

Please only review the fourth commit, based on https://github.com/kubernetes/kubernetes/pull/47345

**Release note**:

```release-note
kubeadm: Make self-hosting work by using DaemonSets and split it out to a phase that can be invoked via the CLI
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @jbeda

### CODE REVIEW

**Comment by timothysc** [comment]
So this stuff needs to be merged in order, so marking this one as do-not-merge till the others are ready.

File: cmd/kubeadm/app/cmd/phases/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/selfhosting.go]: We should follow the conventions of // godoc on any public facing f(n)s.

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: same.

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: I'm not entirely clear what you are doing here and why?

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Could you comment this, it's totally unclear to anyone not working in this area what you are doing, and reference your phases.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Can't we now potentially leave it in a 1/2 deployed state?  and future creates could fail?

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Again comment on why you are loading from the original static manifest.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Do we want to add either a label or name that it was kubeadm that did it?

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: Why even have this?  ¯\_(ツ)_/¯

File: cmd/kubeadm/app/util/apiclient.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient.go]: Shouldn't we poll timeout in case something weird happens on the kubelet?  B/c nothing weird ever happens on the kubelet.

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/util/apiclient.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient.go]: What would you prefer? How long should we wait before timing out?

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: I'll add unit tests shortly

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: I don't know, what do you prefer?

File: cmd/kubeadm/app/util/apiclient.go
Comment (kargakis) [inline cmd/kubeadm/app/util/apiclient.go]: How about a timeout flag in the selfhosted command?

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (GheRivero) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Moved recently to k8s.io/api/core/v1

File: cmd/kubeadm/app/util/apiclient.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient.go]: Can be a future improvement

**Comment by luxas** [comment]
- Rebased
 - Improved the code
 - Commented all functions
 - Added lots of unit tests (in their own commit for easier reviewing) 

In fact, the self-hosting code decreases with this PR with ~80 LoC.
This PR might look scary with a lot of changed lines, but in fact the actual code decreases.

However, the self-hosting code didn't have any comments or unit tests before, but now I've added ~700 LoC of unit tests :).
It's also possible to cleanup more code after this PRs is merged, but saving those cleanup tasks to a future PR

cc @jamiehannaford PTAL

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: nit: lowercase

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Comment that the sequence matters or that it doesn't.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Do we want to tolerate AlreadyExists errors?

File: cmd/kubeadm/app/util/apiclient.go
Comment (kargakis) [inline cmd/kubeadm/app/util/apiclient.go]: Do you prefer adding the /healthz check inside WaitForAPI as a readiness probe inside the static pod or by mutating its spec and then wait for Ready pods here instead of Running?

File: cmd/kubeadm/app/util/apiclient.go
Comment (kargakis) [inline cmd/kubeadm/app/util/apiclient.go]: Can you open an issue so we won't forget? Assign it to me if you like and I should get to it soon.

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: For self-hosted etcd we'll also need to modify the podSpec `command` (mainly the apiserver). I assume we can pass in the additional flags which either overwrite or append to the existing one.

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: 👍  Really like this composable way of mutating the original podSpec

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Do the self-hosted scheduler and controller manager really need to be on the host network? I think we should only use it when absolutely necessary because it uses up port space unnecessarily. If not, can we add a new mutator?

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: s/Use the PodSpec above mentioned PodSpec/Use the above mentioned PodSpec/

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Wait for the Pods to be become Running

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: containers

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: which is suitable

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: can we create a utility function to prepend the prefix? it's a bit brittle doing it manually all the time IMO. I see you added `buildSelfHostedWorkloadLabelQuery` already

File: cmd/kubeadm/app/util/apiclient.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/util/apiclient.go]: is it worth outputting when 0 pods are found? if not, can we move this beneath the `len(apiPods.Items) == 0` check

File: cmd/kubeadm/app/util/apiclient.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/util/apiclient.go]: We might need to disable healthz check on the bootstrap apiserver for self-hosted etcd due to [various race conditions](https://github.com/kubernetes-incubator/bootkube/issues/495), so I think this is the best solution for time being

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: can we change this arg to `0.0.0.0`. that will allow us to take off host network. same for scheduler

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: also, can we make this an empty string? it defaults to pod IP if left blank. I see no reason to hardcode localhost here, especially if we want to move off host network

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: minor nit: I don't think unit tests should be modify external state - whether it be filesystem or DB. But due to the current state of our e2e tests maybe it's better to put here instead. What do you think?

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: I noticed bootkube's [self-hosted controller manager](https://github.com/kubernetes-incubator/bootkube/blob/master/pkg/asset/internal/templates.go#L495) uses `Default` as the DNS policy. @aaronlevy is there any reason for that? Just want to make sure there's no corner cases we're forgetting about.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Can we put all the teardown (L79-87) into a new function and have the init call it after CreateSelfHostedControlPlane? The reason I ask is that it'll need to happen after the etcd-operator deployment and pivot, so doing it now will reduce the future delta

**Review by jamiehannaford — CHANGES_REQUESTED** [review]
Looks great overall, just a few minor issues

File: cmd/kubeadm/app/util/apiclient.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/util/apiclient.go]: We already log _a lot_ of info so being more picky would be nice 😄

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: I don't think so, as I don't recognize any practical purpose for it
The less code the better

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Yes. I have coded some methods for modifying `.command` but didn't include that in this PR for easier reviewing

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: > Do the self-hosted scheduler and controller manager really need to be on the host network

Yes they do, since there is no Pod Network available

> If not, can we add a new mutator?

I can add one to ensure `hostNetwork=true`

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: Hmm, I'm not sure why that would allow us to remove `hostNetwork`...
As I understand this, the current state is "use the host's network, but don't expose anything widely"
We still need hostNetwork for being able to run things without CNI set up.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: I've seen lots of unit tests that create temp files and then remove them.
A quick search for `ioutil.TempFile(` for files `kubernetes/*_test.go` yielded 170 results.
It's a pattern I definitely think we can use

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: I'd rather not do that in this PR. It's set in `app/phases/controlplane/manifests.go` since a long time back.

This is just testing data, but based on actual output from kubeadm.

Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: Also ref this statement: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/

> Here’s one scenario where you would set it. Suppose the Container listens on 127.0.0.1 and the Pod’s hostNetwork field is true. Then host, under httpGet, should be set to 127.0.0.1.

To me it seems like it's required for hostNetwork=true

Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: Yeah, but my point is that the scheduler and controller manager don't need to be on host network. For apiserver (which is on host network), leaving blank will still work because it'll just ping that pod's API

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Wait until the Pods are running ;-)

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Really? I thought pod IPs were addressable without a SDN. I got it working without flannel or anything. Maybe we can shelve this for now then

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Do we want to treat them as separate Pods, or containers as part of the same pod?  

I'm not certain if we ever discussed a tradeoff here?  

The reason I ask is it simplifies the model.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: +1 for mutating function with strings pushed to consts.

**Review by timothysc — CHANGES_REQUESTED** [review]
Please address mine and other comments.  

poke again when done.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Maybe folks would want to scale more apiservers than schedulers/CMs, or vice versa? Just a thought

Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: I think it makes a lot of sense to decouple as three different DaemonSets (==3 different Pods) instead of having one DS and one Pod per master.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting_test.go]: They must be on the host network due to that CNI is uninitialized, so we don't have any other option here really

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: Discussed with @jamiehannaford on Slack and decided to hold off that for now, and re-order the sequence instead. This is fine as-is now and we'll change it if we change it in the future.

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (luxas) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Hmm, I'm not sure what you mean... This works

**Comment by luxas** [comment]
@timothysc @jamiehannaford @kargakis PTAL, updated the PR

**Review by jamiehannaford — APPROVED** [review]
👍

File: cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go
Comment (aaronlevy) [inline cmd/kubeadm/app/phases/selfhosting/podspec_mutation.go]: Looking back at the git blame I believe it had to do with the controller-manager needing to use the dns of the host when cloud-providers were enabled (for example, aws hostnames won't resolve unless you're using the host defined dns server). And at that time `DNSClusterFirstWithHostNet` didn't exist -- but seems like a more reasonable choice now (opened a PR to change in bootkube: https://github.com/kubernetes-incubator/bootkube/pull/629)

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: We may fail on status updates in the DaemonSet so I would make this retriable on 409s. TODO is also fine.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (kargakis) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: failed to update

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
Squashed to two logical comments. Reapplying the labels

**Pull request merged**
